device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java

```diff
@@ -1142,23 +1297,48 @@ private void updateStatus(IotHubConnectionStatus newConnectionStatus, IotHubConn
             //invoke connection status callbacks
             this.log.debug("Invoking connection status callbacks with new status details");
             invokeConnectionStateCallback(newConnectionStatus, reason);
-            invokeConnectionStatusChangeCallback(newConnectionStatus, reason, throwable);
+
+            if (deviceClientConfigs.size() < 2 || newConnectionStatus != IotHubConnectionStatus.CONNECTED)
+            {
+                // When multiplexing, a different method will notify each device-specific callback when that device is online,
+                // but in cases when the tcp connection is lost and everything is disconnected retrying or disconnected, this is where the
+                // callback should be fired
+                invokeConnectionStatusChangeCallback(newConnectionStatus, reason, throwable);
+
+                for (DeviceClientConfig config : deviceClientConfigs)
+                {
+                    deviceConnectionStates.put(config.getDeviceId(), newConnectionStatus);
```

Comment:
This might also be slow for 1000 devices. Thoughts on parallelizing?

Reply:
There isn't a way to speed this up without introducing more threading. This is just putting a key value pair in a hashmap, which shouldn't be much of a bottleneck anyways